### DISCUSSION TITLE
XplatGenerateReleaseNotes : Get only the Pull Request trigger from the execution pipeline

### DISCUSSION THREAD

**Discussion by samyKhezandji** [rfennell/AzurePipelines#1343] [Q&A] [answered]
Hello,

My name is Samy and I am a DevOps engineer in the Data Analytics department of a big French company. I am currently working on a number of DevOps processes and one of my requirements is to automate the creation of release notes in the Azure DevOps Wiki. For this, I opted for the use of the **XplatGenerateReleaseNotes** task, followed by the task **WikiUpdaterTask** (for the publication of release notes) in à Azure Pipeline definition. However, I found that the release notes generated and published in the Azure DevOps Wiki page include all Pull Requests associated with the repository configured for the pipeline in which the tasks are used.

In my use case, I configured the release notes generation task so that on each run a new wiki page is created to contain the Pull Request that triggered the Pipeline as well as the list of associated commits. 

However, I find at each execution of this pipeline a new Wiki page which takes up the elements of the previous page and which add the most recent Pull Requests.

I don't know if this behaviour is normal but my question is : Is there a feature that would allow to get only the Pull Request that triggered my pipeline in each generated Wiki page, without getting the list of all Pull Requests associated with the repository?

This is m configurations of the two tasks **XplatGenerateReleaseNotes** and **WikiUpdaterTask** :

- task: XplatGenerateReleaseNotes@3
          displayName: "Generate release notes"
          inputs:
            outputfile: '$(System.DefaultWorkingDirectory)\releasenotes.md'
            templateLocation: File
            templatefile: ".azuredevops/release-notes-template.md"
            dumpPayloadToConsole: false
            dumpPayloadToFile: false
            replaceFile: True
            getParentsAndChildren: False
            getAllParents: False
            getIndirectPullRequests: False
            stopOnError: True
            considerPartiallySuccessfulReleases: False

- task: WikiUpdaterTask@1
          displayName: "Publish to the wiki"
          inputs:
            repo: "$(System.CollectionUri)$(System.TeamProject)/_git/$(wiki_name)"
            filename: "${{parameters.Database}}/v.$(Build.BuildNumber)-Snowflake-Release-Notes.md" #Note: want a space in the name of the page? Use a '-' here!
            replaceFile: false
            appendToFile: false
            dataIsFile: true
            sourceFile: '$(System.DefaultWorkingDirectory)\releasenotes.md'
            message: "Update from Pipeline"
            gitname: $(Release.RequestedFor)
            gitemail: $(Release.RequestedForEmail)
            useAgentToken: true
            localpath: '$(System.DefaultWorkingDirectory)\repo'

And this is the markdown template file that I created : 

`# 🚀 Release v{{buildDetails.buildNumber}}

## Associated Pull Requests ({{pullRequests.length}})
{{#forEach pullRequests}}
* Pull Request **[#{{this.pullRequestId}} {{this.title}}]({{replace (replace this.url "_apis/git/repositories" "_git") "pullRequests" "pullRequest"}})** 
* Associated Commits (this includes commits on the PR source branch)
{{#forEach this.associatedCommits}}
    - [{{this.comment}}]({{this.remoteUrl}})
{{/forEach}}
{{/forEach}}

## Assets
* **[Release v{{buildDetails.buildNumber}} as ZIP]({{replace buildDetails.project.url "_apis/projects/" ""}}/_apis/git/repositories/{{buildDetails.repository.id}}/items?path=/&versionDescriptor%5BversionOptions%5D=0&versionDescriptor%5BversionType%5D=1&versionDescriptor%5Bversion%5D=uat_v{{buildDetails.buildNumber}}-Merged&resolveLfs=true&%24format=zip&api-version=5.0&download=true)**

---
`
Thanking you in advance for your response and above all thank you for this awesome extension :)

**Comment by rfennell** [comment, marked as answer]
What you are asking for if I have it correct, just the PR that triggered the pipeline run, is actually the original PR functionality I wrote. It was then extended to associated PRs based on community feedback.

Luckily the code is still in the task. So if you look at `ReleaseNotesFunction.ts`  Ln1993 you can see code that gets the triggering PR, but then the associated PRs (found based on comparing PR commit ID to the associated commits for the pipeline) are also added to the same array to be passed to the template.

As I never sort the array, to get the result you want I think you can just take the first PR in the `pullrequests` template object array.

However, we could also consider adding an extra boolean parameter to get only the trigger PR as opposed to any associated PRs. This would have the bonus of making the task run faster as the PR search is fairly slow.

Let me know you thoughts

**Reply by samyKhezandji** [reply]
Thank you for your answer. Fetching the first element of the PRs array is an action that worked for me and returned the desired result. In my work of setting up a CI/CD process including the generation of release notes, the most judicious that I have found for the moment is to associate each release with a wiki page as a release note. This release note contains the PR that triggered the release and then associates this PR with its set of commits. Thus, I can see for each version the PR and the list of commits pointing to the list of modified files.

I think it would be judicious on your side to standardize the obtaining of the PRs associated with the last executions of the pipeline only.

Thanking you 👍

**Reply by rfennell** [reply]
Yes, I think I will add the feature I outlined, thanks for the feedback

**Reply by rfennell** [reply]
See #1344 for tracking

**Reply by rfennell** [reply]
When I checked I found this function was already in place it is the `getPRDetails` this defaults to true and adds PRs that are found by merge commit ID. If it is false then only the triggering PR (if any) is added
